### PR TITLE
State mock

### PR DESCRIPTION
**Pull request by thatch45**
This adds the "mock" option to state.sls and state.highstate. This allows for the state to be run without calling anything. The returned data is the return as if the states all ran successfully.

What would be nice to have here would be the ability to make mock states so that you could program what the mocked returns would look like.

### CODE REVIEW

**Comment by thatch45** [comment]
@jfindlay @whiteinge  let me know if this meets our needs, at the least this is the hard part so you guys should be able to easily take it from here

**Comment by thatch45** [comment]
BTW, this does also report argument warnings

**Comment by whiteinge** [comment]
35 LOC. Game changing addition. This checks out as a bona fide Tom pull req.

**Comment by s0undt3ch** [comment]
Happy New Year!!!!

File: salt/state.py [hunk 1]
```diff
@@ -262,6 +262,23 @@ def ishashable(obj):
...
+    Returns a mocked return dict with inforwation about therun without
+    executing
+    '''
+    # As this is expanded it should be sent into the execution module
+    # layer or it should be turned into a standalone loader system
+    if cdata['args']:
+        name = cdata['args'][0]
+    else:
+        name = cdata['kwargs']['name']
+    return {'name': name,
```
Comment (s0undt3ch) [inline salt/state.py]: We should probably pull the `warnings` key, if present, from cdata.

**Comment by s0undt3ch** [comment]
Perhaps the mock states, considering `<parent dir>/<a dir>/a_state.sls`, should probably live under `<parent dir>/<a dir>/mocks`...

We could probably allow specifying the mock state to use to test different scenarios.

The mock state should have state ids with the same name as the state id being mocked.

... Just suggestions...

**Comment by s0undt3ch** [comment]
Or the mocks loader...

**Comment by s0undt3ch** [comment]
The unitbtest failures seem valid. The call mocks need to be updated....

File: salt/state.py [hunk 2]
```diff
@@ -262,6 +262,23 @@ def ishashable(obj):
     return True
 
 
+def mock_ret(cdata):
+    '''
+    Returns a mocked return dict with inforwation about therun without
```
Comment (cachedout) [inline salt/state.py]: Typos: `information` and `the run`

**Comment by cachedout** [comment]
Yup, this is good. Very good. Very, very good.

File: salt/state.py (hunk 1 above)
Comment (thatch45) [inline salt/state.py]: No need,it is added by the state system to the ret.

File: salt/state.py (hunk 2 above)
Comment (thatch45) [inline salt/state.py]: Fixed

File: salt/state.py (hunk 1 above)
Comment (s0undt3ch) [inline salt/state.py]: Awesome!

**Comment by cachedout** [comment]
Weird. It looks like two copies of the test suite started at the same time. Re-testing.

**Comment by cachedout** [comment]
Go Go Jenkins!

**Comment by thatch45** [comment]
This error has me confused, it makes it look like the pr code is not being called. Maybe it is name overloading with the test suite

**Comment by cachedout** [comment]
It's in the test suite. I'll take care of it today and get this merged. No worries.

**Comment by thatch45** [comment]
oh, I changed the internal options and kept the interface at "mock", so if tests pass from my change we should be good. If not, thanks for hunting it down!

**Comment by cachedout** [comment]
Sure thing. Let's see what the tests say and I'll shepard this one on through. Thanks Tom!

**Comment by thatch45** [comment]
Thank you for putting up with my PRs, I wish I had more time to work on them!

**Comment by rallytime** [comment]
@meggiebot #30185 was backported to include the fix to the test suite from @cachedout. Please see #30189.

**Pull request merged**